miden-crypto/src/merkle/smt/partial.rs

```diff
@@ -273,25 +300,16 @@ impl PartialSmt {
 
         let mut node_hash_at_current_index = leaf.hash();
 
-        // We insert directly into the leaves for two reasons:
-        // - We can directly insert the leaf as it is without having to loop over its entries to
-        //   call Smt::perform_insert.
-        // - If the leaf is SmtLeaf::Empty, we will also insert it, which means this leaf is
-        //   considered tracked by the partial SMT as it is part of the leaves map. When calling
-        //   PartialSmt::insert, this will not error for such empty leaves whose merkle path was
-        //   added, but will error for otherwise non-existent leaves whose paths were not added,
-        //   which is what we want.
         let prev_entries = self
-            .0
             .leaves
             .get(&current_index.value())
             .map(|leaf| leaf.num_entries())
             .unwrap_or(0);
         let current_entries = leaf.num_entries();
-        self.0.leaves.insert(current_index.value(), leaf);
+        self.leaves.insert(current_index.value(), leaf);
```

Comment:
I think it would be good to keep the comment in the second bullet point, or adjust it depending on how we move forward, because how we treat empty leaves is subtle and easy to miss.

Iiuc, we still insert empty leaves into the `Leaves` map, but we _could_ now optimize and skip adding those due to the new tracked logic, so:

```rust
if !leaf.is_empty() {
    self.leaves.insert(current_index.value(), leaf);
}
```

But, I think we generally trade memory for performance, and so it probably makes sense to still keep that "caching"? Any thoughts on this?

In any case, I think it would be good to keep/add a comment here explaining the rationale for dealing with empty leaves.